python/pyarrow/_compute.pyx

```diff
@@ -980,51 +973,45 @@ cdef class _ModeOptions(FunctionOptions):
 
 
 class ModeOptions(_ModeOptions):
-    def __init__(self, n=1, skip_nulls=True, min_count=0):
+    def __init__(self, n=1, *, skip_nulls=True, min_count=0):
         self._set_options(n, skip_nulls, min_count)
 
 
 cdef class _SetLookupOptions(FunctionOptions):
-    cdef:
-        unique_ptr[CDatum] valset
-
     def _set_options(self, value_set, c_bool skip_nulls):
+        cdef unique_ptr[CDatum] valset
         if isinstance(value_set, Array):
-            self.valset.reset(new CDatum((<Array> value_set).sp_array))
+            valset.reset(new CDatum((<Array> value_set).sp_array))
         elif isinstance(value_set, ChunkedArray):
-            self.valset.reset(
-                new CDatum((<ChunkedArray> value_set).sp_chunked_array)
-            )
+            valset.reset(
+                new CDatum((<ChunkedArray> value_set).sp_chunked_array))
         elif isinstance(value_set, Scalar):
-            self.valset.reset(new CDatum((<Scalar> value_set).unwrap()))
+            valset.reset(new CDatum((<Scalar> value_set).unwrap()))
         else:
-            raise ValueError('"{}" is not a valid value_set'.format(value_set))
-
-        self.wrapped.reset(
-            new CSetLookupOptions(deref(self.valset), skip_nulls))
+            _raise_invalid_function_option(value_set, "value set",
+                                           error="type")
```

Comment:
If the type is the problem, then probably `TypeError` should be raised, not `ValueError`.